action/protocol/staking/protocol.go

```diff
@@ -201,6 +204,46 @@ func (p *Protocol) CreateGenesisStates(
 	return errors.Wrap(csm.Commit(), "failed to commit candidate change in CreateGenesisStates")
 }
 
+// CreatePreStates updates state manager
+func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager) (err error) {
+	csm, err := NewCandidateStateManager(sm, true)
+	if err != nil {
+		return err
+	}
+
+	if p.stakingCandidatesBucketsIndexer != nil {
+		return p.handleStakingIndexer(ctx, sm, csm.CandCenter())
+	}
+	return
+}
+
+func (p *Protocol) handleStakingIndexer(ctx context.Context, sm protocol.StateManager, center CandidateCenter) error {
+	rp := rolldpos.MustGetProtocol(protocol.MustGetRegistry(ctx))
+	blkCtx := protocol.MustGetBlockCtx(ctx)
+	epochStartHeight := rp.GetEpochHeight(rp.GetEpochNum(blkCtx.BlockHeight))
+	if epochStartHeight != blkCtx.BlockHeight {
+		return nil
+	}
+	bcCtx := protocol.MustGetBlockchainCtx(ctx)
+	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
+	// only for after fairbank
+	if hu.IsPre(config.Fairbank, epochStartHeight) {
+		return nil
+	}
+
```

Comment:
CreatePreStates() already get the context, so above part can be moved into CreatePreStates()